### ISSUE TITLE
Kanes wait times for 10 minutes trying to save an order because of full table scan of SO_CMNT

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1067]
_Imported from trac ticket 1067.
Created by **kbower**
Opened in trac: Sep 15, 2011, 11:00:50 AM
Last modified in trac: Sep 16, 2011, 9:53:11 PM_
***
**Explanation:**
This is being caused by this type of SQL statement:
```
SELECT ordercomments.orderid AS ordercomments_orderid, ordercomments.sequence AS ordercomments_sequence, ordercomments.orderdate AS ordercomments_orderdate, ordercomments.ordersiteid AS ordercomments_ordersiteid, ordercomments.ordersequence AS ordercomments_ordersequence, ordercomments.commentdate AS ordercomments_commentdate, ordercomments."comment" AS ordercomments_comment, ordercomments.systemgenerated AS ordercomments_systemgenerated, ordercomments.regarding AS ordercomments_regarding, ordercomments.username AS ordercomments_username
FROM ordercomments
WHERE ordercomments.orderid IS NULL AND ordercomments.sequence = 4 AND ordercomments.orderdate IS NULL AND ordercomments.ordersiteid IS NULL AND ordercomments.ordersequence IS NULL
```(full table scan of 20 million records)
Sqlalchemy is trying to merge !OrderComment transfer object and since needed to tell sqlalchemy that `allow_partial_pks=True` (because gers has no pk, only a unique constraint with includes a column which is generally null), sqlalchemy thinks it needs to go to the database to check whether the comment exists.

----

Ok, I just got off the phone with a store that's been waiting on an order to save for approx 10min now. A yellow message box briefly appears in the upper-right corner of the screen over the scrolling "Saving Order" button stating something about 2/3 attempts to retry after a communication failure. The order was 0914118AH96.

Unfortunately I wasn't able to catch the entire error message because it disappeared quickly and my remote desktop isn't instantaneous.

I can't find anything wrong with the network at the moment that could cause this kind of problem to occur. Ping times to the server are ~43ms and ~25ms to the store, and my remote control to that PC isn't giving me any problems.

Could this happen if the client was stuck waiting on a response from the server? 

Let me know.
Alex

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 11:01:40 AM:_

Yes, what that means is that a call to the server has failed (timed-out, refused, network dropped, any number of reasons why) and what we did was put in a framework to sort of just "try it again" in most cases but notify the user why it was taking so long (thus the yellow warning box). After 3 attempts it just gives up completely. Is this is only occurrence of this that you know of or is this happening with regularity?

PS this is likely the old "I/O error" that we used to see now and again but the point of this new framework is to instead of just blowing up it tries to gracefully recover as best it can.

Rob

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 11:01:54 AM:_

The way they're describing it to me, it sounds like it's happening pretty often. Unfortunately I can't verify this personally every time, and when I get to them after the fact, everything looks fine. I did manage to catch this one in the act and it was still going for at least 10 minutes before they gave up and keyed the order in GERS to get the customer taken care of.

Is there a way to see from the server side when connections drop or are retried by the client?

Alex

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 11:02:18 AM:_

I also wish I had a better idea of how often this is happening.  I have three suspicions:

1) This really was a network issue.  (That doesn't sound like it is the case from your last email, but haven't ruled it out.)
2) This is an ill side effect of how I fixed the speed issue from 2 days ago.  It is quite feasible that certain queries are extremely complicated for oracle now and are taking forever.
3) With the newer release we are loading more information and processing more and each process for the web server probably needs more RAM than before.  Collectively this may be translating to a need for more RAM.

3) is most easily solved but would have some financial implications for Kane's monthly VM bill.
2) is my most likely suspect, so, at the risk of filling the disk (I'll make a cron) and slightly slowing down the server, I'm going to turn on a bit of logging this morning, unless stores have already opened (when do they open?)

Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 11:02:34 AM:_

I spoke to the office manager who I was working on this with yesterday and I suggested that they try and log in another session from the same PC the next time they had a sale get stuck on the save. 

She called me back today with some interesting info. As soon as they logged in on another browser window, the sale saved immediately. I've asked them to try it again and let me know if they get the same result to try and rule out a possible coincidence. 

I don't know what this means on the back-end of things, since I'm not that familiar with the way the application works yet, but I hope this is useful for troubleshooting.

Alex

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 11:06:27 AM:_

That makes me question if that PC is having network issues?

Also helpful will be if they write down what the yellow warning says.  And if you can note the time and order for which it is happening, I've got a bunch of logging turned on today.

**Comment by kbower** [comment]
_**alex** trac comment on Sep 15, 2011, 11:09:35 AM:_

That yellow message box only appears for a few seconds. The store didn't actually mention it to me at all, I noticed it while I had one of their computers remote controlled. Is there a way to make it persist for the duration of the retry?

I initially suspected the network too, but the fact that my remote control works just fine and the second browser window is able to connect without any delays makes be wonder...

From: Product Development@RetailArchitects [mailto:issues@retailarchitects.com]
Sent: Thursday, September 15, 2011 11:06
Subject: Ticket #1067, Status=new: Kanes wait times for 10 minutes trying to save an order

Ticket #1067 (new support)
Kanes wait times for 10 minutes trying to save an order
Owner:

None

Reporter:

kbower

Milestone:

None

Priority:

critical

Severity:

None

Comments: (by kbower)
That makes me question if that PC is having network issues? Also helpful will be if they write down what the yellow warning says. And if you can note the time and order for which it is happening, I've got a bunch of logging turned on today. On 9/15/2011 10:46 AM, Alexander Harvey wrote:

________________________________

Ticket URL: https://trac.retailarchitects.com:444/trac/ticket/1067
trac https://trac.retailarchitects.com:444/trac
RMS

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 2:25:34 PM:_

Ok, I just got another report from a different store for this same issue. The manager there noticed that all of the orders affected were FPUs.

Here's the e-mail he sent me:
--------------------------
All orders written in the new POS as FPU using store inventory are taking about 3-5 minutes to save since this week's update.  The orders listed below are examples of orders that I've seen this on in the last two days.

0913117AC74
0913117AC71
0914117AH65
0914117AG97

Please let me know of any questions you may have.

Thanks,
Dewey
--------------------------

Hopefully this is helpful.
-Alex

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 2:31:44 PM:_

If it's FPU, it isn't FPU alone.  I just saved one for TRAIN and it didn't take very long.

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 15, 2011, 5:02:09 PM:_

-------- Original Message --------
Subject:         RE: #1067: ten minute wait time
Date:         Thu, 15 Sep 2011 20:47:48 +0000
From:         Alexander Harvey <aharve@kanesfurniture.com>
To:         Kent Bower <kb@retailarchitects.com>
CC:         Rob Neville <rn@retailarchitects.com>, Wayne Liburd <wliburd@kanesfurniture.com>, dev <dev@retailarchitects.com>



I've got another example from Store 17. This one was just moments ago and took approximately 4 minutes to save.

0915117AM41

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 6:49:12 AM:_

Ok, I'm ssh'ing from the iPhone, but this is interesting.... According to these time stamps, the actual requests only took about a second each. 

(I assume you would have mentioned it if you had gotten a yellow warning)

Two theories:  the disk starts thrashing page faults (we know this is going on, we just don't know the effect it has). 
Or if there aren't enough threads, your request is waiting in queue for a slow one before yours to finish.  This is quite possible. 

tg2env)[rarch@kanes httpd]$ grep group/save error_log
[Thu Sep 15 17:02:52 2011] [error] 17:02:52,875 INFO  [auth] -- repoze.who request started (/group/save) --
[Thu Sep 15 17:02:54 2011] [error] 17:02:54,009 INFO  [auth] -- repoze.who request ended (/group/save) --
[Thu Sep 15 17:04:12 2011] [error] 17:04:12,284 INFO  [auth] -- repoze.who request started (/group/save) --
[Thu Sep 15 17:04:13 2011] [error] 17:04:13,512 INFO  [auth] -- repoze.who request ended (/group/save) --
(tg2env)[rarch@kanes httpd]$ grep "INSERT INTO grouppermissions" error_log
[Thu Sep 15 17:02:53 2011] [error] 17:02:53,558 INFO  [sqlalchemy.engine.base.Engine.0x...6fd0] INSERT INTO grouppermissions (groupname, permissionname) VALUES (:groupname, :permissionname)
[Thu Sep 15 17:04:12 2011] [error] 17:04:12,939 INFO  [sqlalchemy.engine.base.Engine.0x...6fd0] INSERT INTO grouppermissions (groupname, permissionname) VALUES (:groupname, :permissionname)
(tg2env)[rarch@kanes httpd

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 6:51:26 AM:_

comment:9 was in reply to Rob:
**did you just bounce apache?** If not, then the problem is systemic because I got a 1-2 minute pause trying to save a user group just now...

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 9:37:19 AM:_

Good point.  On the other hand, we partly didn't even want the thing to show up at all in the first place because generally the user shouldn't care: if there is a network glitch and the request is retried, who cares?  So I hesitate adding this as an annoyance that someone needs to close, you know?

Rob and I are starting surmise that this might have to do with too few threads for processing all the requests or too little RAM (which is related to the not being able to set up more threads), or too little RAM causing disk pagination possibly.

(Our app may require more RAM than we had estimated.)

Another thought I just had: Wayne, can you check the oracle logs for deadlocks?  We'll probably get some, but my concern would be if the number of deadlocks has increased recently.  (Actually, the store would have probably reported if they saw deadlocks since it would show up as an error, but its a thought.)

We still don't have a good feel for how often this is occurring.   Would you say it is every store and maybe a dozen times a day for each store or is it rarer than that?

/
On 9/16/2011 9:12 AM, Alexander Harvey wrote:
/
That would be more reasonable if the yellow communications error was persistent. Many of the people I've spoken to about this at our stores never noticed it at all until I took control of their PC and pointed it out when it appeared.
/
Kent Bower <kb@retailarchitects.com> wrote: /
The cursor is probably unrelated and probably just spinning because it was never told to stop spinning. Please find out whether they got a yellow communications warning prior to the cursor spinning.

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 9:58:48 AM:_

Keyed a few transaction in Train this morning to see how long it would take to save. Here are the transaction and the save time. These were keyed at Kanes corporate, and our network here may be faster than the lines at the stores.
 
Tim
 
0916101NV81            Save time approx 14 seconds
0916101NV82            Save time approx 15 seconds
0916101NV83            Save time approx 25 seconds (Substitutions and EFD discount)
0916101NV84            Save time approx 65 seconds  Tax exempt and TYD Discount
0916101NV85            Save time approx 19 seconds

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 11:49:05 AM:_

Found it.
Won't bore you with the details about the lack of GERS primary key for the so_cmnt and the long story of how this problem came about, but, thank you, Tim: after looking at the orders you created, I found "the" problem (unless there is more than one cause) and recreated an order that took about 4 minutes to save.

I'll start looking into a solution, but you can tell the managers that, indirectly, it has to do with overrides and manager approvals being written as comments.  The more of those you have, the worse this will be.

Actually, I suppose there is a chance Daniel would like to keep it the way it is since it punishes people who save orders needing manager approval.  ;)

Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 16, 2011, 9:53:11 PM:_

(Changeset [[1716]](../commit/21a40b00dc6dc946963b783aa980aedf198a0d2a))
 * special case for OrderComment to avoid searching for a partial pk if we know it can't exist.  Closes #1067 (Kanes wait times for 10 minutes trying to save an order because of full table scan of SO_CMNT)

**Comment by kbower** [comment]
_**kbower** trac status change on Sep 16, 2011, 9:53:11 PM:_ **closed**
